syntax = "proto3";

package gaiachain;

import "agent.proto";
import "entity.proto";

message SCPayload {
    enum Action {
        CREATE_AGENT = 0;
        CREATE_PACKAGE = 1;
        CREATE_ENTITY_BATCH = 2;
        MOVE_ENTITY_BATCH = 3;
        UPDATE_PACKAGE = 4;
        CREATE_REPLANTATION = 5;
    }

    Action action = 1;

    // Approximately when transaction was submitted, as a Unix UTC
    // timestamp
    uint64 timestamp = 2;

    // The transaction handler will read from just one of these fields
    // according to the Action.
    CreateAgentAction create_agent = 3;
    CreatePackageAction create_package = 4;
    CreateEntityBatchAction create_entity_batch = 5;
    MoveEntityBatchAction move_entity_batch = 6;
    UpdatePackageAction update_package = 7;
    CreateReplantationAction create_replantation = 8;
}


message CreateAgentAction {
    Agent agent = 1;
}

message CreatePackageAction {
    Package package = 1;
}

message UpdatePackageAction {
    string id = 1;
    Entity entity = 2;
}

message CreateEntityBatchAction {
    repeated Entity entities = 1;
}

message MoveEntityBatchAction {
    string entity_batch_id = 1;
    //EntityBatch.Status status = 2;

    // step 2
    uint64 breaking_date = 3;
    uint64 end_fermentation_date = 4;
    uint64 beans_volume = 5;

    // step 3
    uint64 reception_date = 6;
    uint64 transport_date = 7;
    string buyer = 8;
}

message CreateReplantationAction {
    Replantation replantation = 1;
}